### PR TITLE
[specific ci=3-04-Docker-Compose-Fixtures] Add pull-by-digest support for schema 1 images

### LINKED ISSUE #1186: Support pull by image digest
Currently imagec is only able to pull images by tag.  For example:

```
docker pull busybox:1.24.2
```

We need to provide support for pulling images via image digest:

```
docker pull busybox@sha256:4a731fb46adc5cefe3ae374a8b6020fc1b6ad667a279647766e9a3cd89f6fa92
```

While vic (via imagec) will attempt a pull when a user provides a digest, it's not pulling via the digest -- it's actually pulling via the latest tag.  Since the digest is ignored the wrong image might be pulled.

Depends on #1484

### PR DESCRIPTION
**Pull request by anchal-agrawal**
Fixes #1186

Follow-up issue for schema 2: #5187

### CODE REVIEW

File: lib/imagec/docker.go [hunk 1]
```diff
@@ -335,8 +349,12 @@ func FetchImageManifest(ctx context.Context, options Options, schemaVersion int,
 	if err != nil {
 		return nil, "", err
 	}
-	url.Path = path.Join(url.Path, options.Image, "manifests", options.Tag)
 
+	manifestRef := options.Tag
+	if digested, ok := options.Reference.(reference.Canonical); ok {
```
Comment (jzt) [inline lib/imagec/docker.go]: Seems a little D.R.Y... Any reason why you can't just call `manifestRef := tagOrDigest(options.Reference, options.Tag)`?

Comment (anchal-agrawal) [inline lib/imagec/docker.go]: Oh, I forgot to change that. Good catch!

**Comment by cgtexmex** [comment]
LGTM

I'll echo @jzt -- Nice work!

File: lib/imagec/docker.go [hunk 2]
```diff
@@ -455,6 +481,15 @@ func getManifestDigest(content []byte) (string, error) {
 	}
 
 	log.Debugf("Canonical Bytes: %d", len(bytes))
+
+	// Verify the manifest digest if the image is pulled by digest.
+	// https://docs.docker.com/registry/spec/api/#content-digests
+	if digested, ok := ref.(reference.Canonical); ok {
```
Comment (fdawg4l) [inline lib/imagec/docker.go]: Can you add a comment why line `493` is ok when this type assertion fails?

File: lib/imagec/docker.go [hunk 3]
```diff
@@ -416,6 +426,22 @@ func decodeManifestSchema1(filename string, options Options, registry string) (i
...
+	if !verifier.Verified() {
+		return fmt.Errorf("image manifest verification failed for digest %s", digested.Digest())
+	}
+
+	return nil
+}
+
 // decodeManifestSchema2() reads a manifest schema 2 and creates a Docker
 // defined Manifest structure and returns the digest of the manifest as a string.
 func decodeManifestSchema2(filename string, options Options) (interface{}, string, error) {
```
Comment (fdawg4l) [inline lib/imagec/docker.go]: It's not strictly part of this change, but this feels like it should be part of an `interface` where schema1 and schema2 are implementations.

File: lib/imagec/docker.go (hunk 2 above)
Comment (fdawg4l) [inline lib/imagec/docker.go]: Also is there a test that exercises the !ok path?

File: lib/imagec/imagec.go [hunk 4]
```diff
@@ -158,29 +155,21 @@ func init() {
...
-	}
-
-	ic.Tag = reference.DefaultTag
-	if !reference.IsNameOnly(ref) {
-		if tagged, ok := ref.(reference.NamedTagged); ok {
+func (ic *ImageC) ParseReference() {
+	if reference.IsNameOnly(ic.Reference) {
+		ic.Tag = reference.DefaultTag
+	} else {
+		if tagged, isTagged := ic.Reference.(reference.NamedTagged); isTagged {
```
Comment (fdawg4l) [inline lib/imagec/imagec.go]: Minor. I think the idiom is to use `ok` and test for `!ok`.

Comment (anchal-agrawal) [inline lib/imagec/imagec.go]: I see. Here, we want to set `ic.Tag` if the type-assert is successful, so we may want to keep this as-is.

File: lib/imagec/docker.go (hunk 2 above)
Comment (anchal-agrawal) [inline lib/imagec/docker.go]: Done and done! Please see d226c7c.

**Pull request merged**